File: vizque/vizque.py
import requests
import xml.etree.ElementTree as ET
from graphviz import Digraph
import shutil
import os


# 取得したファイルはtmpdirに生成
# サーバーからローカルホストでブラウザに表示するように変更


def to_get_suggest_word(query: str) -> list:
    r = requests.get(
        f'http://www.google.com/complete/search?hl=en&q={query}&output=toolbar')
    xml = r.text  # xmlレスポンスの取得
    root = ET.fromstring(xml)  # xmlデータを読み込みます
    suggest_word = [
        suggest.attrib["data"]
        for CompleteSuggestion in root
        for suggest in CompleteSuggestion
    ]
    return suggest_word


class CreateDGnode:
    def __init__(self):

        self.dg = Digraph(format="svg", engine='fdp')

    def add_node(self, query, suggest_word):
        # 有向グラフ
        self.dg.node(query)
        for sug in suggest_word[1:]:
            self.dg.node(sug)
            self.dg.edge(query, sug)

    def output(self):

        import re
        id = 0
        node_orig = []

        pr = re.compile("->")
        tmp_node= [] #nodes data
        node_data_words = {}
        for i in self.dg:
            if i == "digraph {" or i == "}":
                continue
            w_i = i.replace("\t", "").replace("\"", "")
            result = pr.search(w_i)
            if result is None:
                n_dict = { "id": str(id), "word": w_i, "url": f"https://www.google.com/search?q={w_i}"}
                tmp_node.append(n_dict) #data
                node_data_words[w_i] = str(id) # word(key): id(val) 
                id += 1
            else:
                node_orig.append(w_i.split(" -> "))
        

        for k in node_data_words.keys():
            for i, orig in enumerate(node_orig):
                from_ = orig[0]
                to_ = orig[1]
                if k == from_:
                    node_orig[i][0] = node_data_words[k]
                if k == to_:
                    node_orig[i][1] = node_data_words[k]
  

        def to_edge(edge):
            direction = {"from": edge[0], "to": edge[1]}
            return direction

        edge_data = [obj for obj in list(map(to_edge, node_orig))] #edge(from -> to)

        vis_grapf = {"nodes": tmp_node, "edges": edge_data}

        import json
        with open("./jsonData/grapf.json", mode="w") as fn:
            json.dump(vis_grapf, fn, indent=5, ensure_ascii=False)


def main():
    query = input("Please enter to search word: ")
    suggest_word = to_get_suggest_word(query)  # xmlレスポンスからデータの取得
    dg = CreateDGnode()

    for _ in range(10):
        for query in suggest_word:
            suggest_word = to_get_suggest_word(query)
            dg.add_node(query, suggest_word)

    dg.output()

File: vizque/vizque.nim
# nim c -o:output -r vizque.nim
import
  httpClient,
  strtabs,
  re,
  xmlparser,
  xmltree,
  tables,
  json


type
  Json = object
  QueryDatas = Table[string, string]
  SecTables = seq[QueryDatas]

type
  GraphDatas = ref object
    nodes: SecTables
    edges: SecTables


var
  node: QueryDatas = initTable[string, string]()
  node_init = node
  edge: QueryDatas = initTable[string, string]()
  edge_init = edge
  graph = initTable[string, SecTables]()
  # querylist: seq[string] = newSeq[string]()
  #Node
  ci:int  = 0
  id: int = 0
  #edge
  tmp_from_id = 0
  from_id: int = 0
  to_id: int = 0


#検索ワードの取得
# echo "Please enter to search word: "
# let reader =  readLine(stdin)


#オブジェクト定義
var G:GraphDatas = new GraphDatas

proc create_node(query: string, id: int) =
  #queryからnodeを作成する

  #ノードの作成
  node["id"] = $id
  node["label"] = query
  node["url"] = "https://www.google.com/search?q=${query}" % {"query": query}.newStringTable
  #nodesに格納
  G.nodes.add(node)
  #ノードの初期化
  node = node_init


proc create_edge(from_id: int, to_id: int) =
  #queryからedgeを作成する
  edge["from"] = $from_id
  edge["to"] = $to_id
  #edgesに格納
  G.edges.add(edge)
  #edgeの初期化
  edge = edge_init


proc outputGraphData(nodes: SecTables, edges: SecTables) =
  #jsonに結果を出力する
  graph["nodes"] = nodes
  graph["edges"] = edges

  block:
    #json用オブジェクトの作成
    let to_json = %* graph
    let f = open("../react-app/jsonData/graph.json", FileMode.fmWrite)
    #jsonファイルに書き込み
    f.write(to_json.pretty(indent=5))
    f.close()
      




proc querygetter(query: string): seq[string] =
  #検索結果を格納するsequence
  var searchQuery = newSeq[string]()
  #空白を埋めてget用URLの作成
  var rep_query = query.replace(re"\s+", "%20")
  var url = "http://www.google.com/complete/search?hl=en&q=${query}&output=toolbar" % {"query": rep_query}.newStringTable

  # httpクライエントの作成
  let client = newHttpClient()
  let response = client.request(url)
  #bodyを取得
  var body: string = response.body
  #xmlをパース
  let xmls = parseXml(body)
  #suggestion タグの取得
  var tag_suggestion = xmls.findAll("suggestion")
  tag_suggestion = tag_suggestion[1..^1] #頭はqueryなので含まない
  for tag in tag_suggestion:
    inc(id)
    #tagからdata属性(関連キーワード)を取得
    let relationQuery = tag.attr("data")
    #relationqueryからnodeを作成
    create_node(relationQuery, id)
    #tmp_from_idをfrom_id, idをto_iとしてedgeを作成
    create_edge(tmp_from_id, id)

    searchQuery.add(relationQuery)
  #jsonに出力
  outputGraphData(G.nodes, G.edges)

  return searchQuery #検索クエリ候補を返す


# test
var reader = "whale shark"
create_node(reader, id)
#from側のidをセット
tmp_from_id = id


var querys: seq[string] = querygetter(reader)

echo querys









